### ISSUE TITLE
save -d . subds/subsubds takes forever for no good reason

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad#5521]
might be the same underlying issue as of #5283 (didn't analyze details yet)

the intention from running this command, without `-r`, is just to commit as possibly updated state of the sub/subds within hierarchy.  In majority of the cases it might not even be changed in my use case.  But it takes really forever IMHO for no valid reason

```shell
$> datalad --version
datalad 0.14.0.dev46

$> time datalad save -d . -J1  -m 'Fast forward psychoinformatics-de/hirni-toolbox recursively' psychoinformatics-de/
add(ok): psychoinformatics-de (file)
Total (1 skipped):  50%|████████████████████▌                    | 1.00/2.00 [00:00<00:00, 806 datasets/s]
.... was stuck at this already for the duration of me opening new tab, and composing this issue ... will later report output of time
```
NB rate of progress (806 datasets/s is really "funny" given it been stuck for so long having accomplished virtually nothing ;) )

oh -- finally it is "done" with a considerable effort on user and system:

```shell
$> time datalad save -d . -J1  -m 'Fast forward psychoinformatics-de/hirni-toolbox recursively' psychoinformatics-de/
add(ok): psychoinformatics-de (file)
action summary:                                                                                           
  add (ok: 1)
  save (notneeded: 2)
datalad save -d . -J1 -m  psychoinformatics-de/  49.94s user 102.12s system 52% cpu 4:51.77 total
```

**Comment by yarikoptic** [comment]
<details>
<summary>some observations/notes</summary> 

the culprit is that at the end it does run
```
[DEBUG  ] Async run:
|  cwd=/mnt/btrfs/datasets/datalad/crawl
|  cmd=['git', '-c', 'diff.ignoreSubmodules=none', 'commit', '-m', 'Fast forward psychoinformatics-de/hirni-toolbox recursively', '--']
[DEBUG  ] Launching process ['git', '-c', 'diff.ignoreSubmodules=none', 'commit', '-m', 'Fast forward psychoinformatics-de/hirni-toolbox recursively', '--']
[DEBUG  ] Process 8095 started
```
i.e. commit without specifying any path, so as if nothing to be committed (there is "no staging area" for datalad, right?) which takes that while (>2 minutes cold, >1 warm)...

trying to debug why we are saving and got to
```
> /home/yoh/proj/datalad/datalad-master/datalad/support/gitrepo.py(3593)save_()
-> amend = kwargs.get('amend', False)
(Pdb) p paths
None
(Pdb) l
3588            """Like `save()` but working as a generator."""
3589            from datalad.interface.results import get_status_dict
3590    
3591            status = self._save_pre(paths, _status, **kwargs)
3592            import pdb; pdb.set_trace()
3593 ->         amend = kwargs.get('amend', False)
3594            if not status and not (message and amend):
3595                # all clean, nothing todo
3596                lgr.debug('Nothing to save in %r, exiting early', self)
3597                return
3598    
(Pdb) p status
OrderedDict([(PosixPath('/mnt/btrfs/datasets/datalad/crawl/psychoinformatics-de'), {'state': 'untracked', 'type': 'dataset'})])
(Pdb) p kwargs
{'git': None, 'untracked': 'no', 'amend': False}
(Pdb) p _status
{PosixPath('/mnt/btrfs/datasets/datalad/crawl/psychoinformatics-de'): {'state': 'untracked', 'type': 'dataset'}}
```
why is it "untracked"??  but if I go up in traceback - gets even more interesting
```
301                 if not all(p['state'] == 'clean' for p in pds_status.values()) or \
302                         (amend and message):
303                     for res in pds_repo.save_(
304                             message=message,
305                             # make sure to have the `path` arg be None, as we want
(Pdb) 
306                             # to prevent and bypass any additional repo.status()
307                             # calls
308                             paths=None,
309                             # prevent whining of GitRepo
310                             git=True if not hasattr(ds.repo, 'annexstatus')
311                             else to_git,
312                             # we are supplying the full status already, do not
313                             # detect anything else
314                             untracked='no',
315                             _status=pds_status,
316  ->                         amend=amend):
```
so `path` is intentionally left at `None`... hm... will dig more later

</details>

**Comment by mih** [comment]
OK, after failing to shed light on https://github.com/datalad/datalad/pull/5576 let me try it here:

> so path is intentionally left at None... hm... will dig more later

This is done, because we pass `_status`. This will be used, instead of triggering another status run inside -- we already have the result and we can just pass it on.

> why is it "untracked"??

It is `untracked`, because `save()` prepares this fake record for any dataset in the trace between a to-be-saved path and the root dataset. It triggers the a call to `GitRepo._save_add_submodules()`

Which leaves:

> i.e. commit without specifying any path, so as if nothing to be committed (there is "no staging area" for datalad, right?) which takes that while (>2 minutes cold, >1 warm)...

The statement "no staging area for datalad" is not correct. It is very much aware of a staging area. It even makes sure to keep staged things staged, when something else is to be committed (grep for `need_partial_commit`). "not staging area for datalad" simply means that datalad's own workflows do not require a staging area to be distinguished -- not that it is incompatible with the concept.

I will now look for the rational behind a `commit` without a path specification.

**Comment by mih** [comment]
OK, so the reason for `git commit --` -- which presumably is the key issue here, is the following condiiton

https://github.com/datalad/datalad/blob/76314749582b79a279baac9fc5cbf84c5eaa4ad5/datalad/support/gitrepo.py#L3318

If that is forced to `True`, the call would end up listing files `['<sub>', '.gitmodules']` after `--`.

My memory, assisted by `git blame` proclaims that this is a straight take over from datalad-revolution times. I believe the motivation was to cut down on processing time for monstrous cmdlines. They can happen when the `status` report was complicated and many files have changes. `GitRepo.save_()` stages everything before the commit, so the commit call can be without file constraints -- unless something was staged before. This is reflected in the present conditional.

Now clearly, this has unacknowledged side-effects. Working out a better `need_partial_commit` condition is the key issue here IMHO.

**Comment by yarikoptic** [comment]
> > why is it "untracked"??
> 
> It is `untracked`, because `save()` prepares this fake record for any dataset in the trace between a to-be-saved path and the root dataset. It triggers the a call to `GitRepo._save_add_submodules()`

FWIW IIRC anything correctly about this issue this far in the future:  above explanation might be providing "rationale" but IMHO the value "untracked" is incorrect since that `subds` is tracked!  And such "assign an incorrect value" IIRC then boils down the issue that for "untracked" there is more logic needed to figure out the status of the `subds`, either it conflicts with any paths within super's `ds` (and thus needing some not really needed "listing" for the superds IIRC) and not just `subds/subsubds`.

That is why I took that approach in https://github.com/datalad/datalad/pull/5576 although here and there @mih identified `need_partial_commit` as the "right nob to turn" ;)

**Comment by yarikoptic** [comment]
the issue might have been resolved! I used datalad instead 0.15.6 I used in that env to current master and even then added a change to that sub/subdatset and it worked out just fine and quick!

```
$> time git -C psychoinformatics-de status
On branch master
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   hirni-toolbox (new commits)

no changes added to commit (use "git add" and/or "git commit -a")

real    0m0.012s
user    0m0.000s
sys     0m0.012s
(dev3) (git)/mnt/btrfs/datasets/datalad/crawl:[master]
#43 !378 [0].....................................:Wed Aug 24 12:14:22:.
smaug:/mnt/btrfs/datasets/datalad/crawl
$> time datalad save -d . -m 'TEMP: updating hirni-toolbox after a dummy change!' psychoinformatics-de/hirni-toolbox
add(ok): hirni-toolbox (dataset)                                                                                                                                                                                     
add(ok): .gitmodules (file)                                                                                                                                                                                          
save(ok): psychoinformatics-de (dataset)                                                                                                                                                                             
add(ok): psychoinformatics-de (dataset)                                                                                                                                                                              
add(ok): .gitmodules (file)                                                                                                                                                                                          
save(ok): . (dataset)                                                                                                                                                                                                
action summary:                                                                                                                                                                                                      
  add (ok: 4)
  save (ok: 2)

real    0m1.848s
user    0m1.188s
sys     0m0.775s
(dev3) (git)/mnt/btrfs/datasets/datalad/crawl:[master]
#44 !379 [0].....................................:Wed Aug 24 12:14:37:.
smaug:/mnt/btrfs/datasets/datalad/crawl
$> datalad --version
datalad 0.17.3+34.gdc77a088f

```
